File: v1/Java/boj/Beginner/C_BinarySearch/Boj_2805.java
package boj.Beginner.C_BinarySearch;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.StringTokenizer;

public class Boj_2805 {
    static StringTokenizer st;
    static final BufferedReader br =
            new BufferedReader(new InputStreamReader(System.in));
    static List<Long> arr;

    private static long solve(int m) {
        long left = 0;
        long right = Collections.max(arr);
        long ans = 0;

        while (left <= right) {
            long mid = (left + right) / 2;
            long tmp = 0;

            for (long a : arr)
                if (a > mid)
                    tmp += a - mid;

            if (tmp < m) right = mid - 1;
            else {
                left = mid + 1;
                ans = Math.max(ans, mid);
            }
        }

        return ans;
    }

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        arr = new ArrayList<>();
        st = new StringTokenizer(br.readLine());

        for (int i = 0; i < n; i++)
            arr.add(Long.parseLong(st.nextToken()));

        System.out.println(solve(m));
    }
}

File: Kotlin/src/main/kotlin/beginner/c_BinarySearch/boj_2805.kt
package beginner.c_BinarySearch

import java.io.BufferedReader
import java.io.InputStreamReader

private fun solve(arr: LongArray, m: Int): Long {
    var start: Long = 0
    var end: Long? = arr.maxOrNull()

    while (start < end!!) {
        val mid: Long = (start + end) / 2
        var tmp: Long = 0

        for (a in arr)
            if (a > mid)
                tmp += a - mid

        if (tmp < m) end = mid
        else start = mid + 1
    }

    return end - 1
}

fun boj_2805() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val (n: Int, m: Int) = readLine()
        .split(" ")
        .map { it.toInt() }

    val arr: LongArray = readLine()
        .split(" ")
        .map { it.toLong() }
        .toLongArray()

    /*
    val arr = emptyList<Int>().toMutableList();
    for (i in 0 until n)
        arr.add(readLine()!!.toInt())
    */

    println(solve(arr, m))
}